cpp/src/spatial/hausdorff.cu

```diff
@@ -0,0 +1,178 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+#include <type_traits>
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cuspatial/error.hpp>
+
+namespace {
+
+// const uint32_t NUM_THREADS = 1024;
+
+template<typename T>
+constexpr auto magnitude_squared(T a, T b) {
+    return a * a + b * b;
+}
+
+template <typename T>
+__global__ void kernel_hausdorff_full(int trajectory_count,
+                                      T* xs,
+                                      T* ys,
+                                      uint32_t* pos,
+                                      T* results)
+{
+    int block_idx = blockIdx.y * gridDim.x + blockIdx.x;
+
+    if (block_idx < trajectory_count * trajectory_count)
+    {
+        int a_idx   = block_idx % trajectory_count;
+        int a_begin = a_idx == 0 ? 0 : pos[a_idx - 1];
+        int a_end   =                  pos[a_idx];
+
+        int b_idx   = block_idx / trajectory_count;
+        int b_begin = b_idx == 0 ? 0 : pos[b_idx - 1];
+        int b_end   =                  pos[b_idx];
+
+        T min_dist_sqrd = 1e20;
+
+        int max_threads = b_end - b_begin; // why b vs a? minimum length of the two, I guess?
+
+        if (threadIdx.x < max_threads)
+        {
+            T bx = xs[b_begin + threadIdx.x];
+            T by = ys[b_begin + threadIdx.x];
+
+            for (int i = a_begin; i < a_end; i++)
+            {
+                T ax = xs[i];
+                T ay = ys[i];
+                T dist_sqrd = magnitude_squared(bx - ax, by - ay);
+                min_dist_sqrd = min(min_dist_sqrd, dist_sqrd);
+            }
+        }
+
+        if (min_dist_sqrd > 1e10) // this may make some of the following statements uncessary.
+        {
+            min_dist_sqrd = -1;
+        }
+
+        __shared__ T dist_sqrd_shared[1024];
+
+        dist_sqrd_shared[threadIdx.x] = -1;
+
+        __syncthreads(); // not sure if this is necessary
+
+        if (threadIdx.x < max_threads) // not sure why this condition exists.
+        {
+            dist_sqrd_shared[threadIdx.x] = min_dist_sqrd;
+        }
+
+        __syncthreads(); // not sure if this is necessary
+
+        for (int offset = blockDim.x / 2; offset > 0; offset >>= 1)
+        {
+            if (threadIdx.x < offset)
+            {
+                dist_sqrd_shared[threadIdx.x] = max(dist_sqrd_shared[threadIdx.x],
+                                                    dist_sqrd_shared[threadIdx.x + offset]);
+            }
+
+            __syncthreads(); // in a loop? yikes!
```

Comment:
This is needed when you do a reduction. The # of loops is log(blockDim.x) which is no more than 10.